File: mssp/examples/018-who-the-declaration-serves/src/DMS/report.mjs
// What a person is shown.
//
// An incentive is never printed as a single number. It is printed as the pair
// it was computed from, because "+3" without "3 -> 6" is a claim nobody can
// re-derive.
const pad = (v, w) => String(v).padEnd(w);

export function rows(result) {
  const lines = ["  source        held   kept   declared                note"];
  for (const row of result.rows) {
    lines.push(`  ${pad(row.source, 13)} ${pad(row.held, 6)} ${pad(row.kept.length, 6)} `
      + `${pad(row.incomplete_because ?? "-", 23)} ${row.note ?? ""}`.trimEnd());
  }
  lines.push(`  total kept: ${result.total}`);
  return lines.join("\n");
}

export function incentives(measured) {
  const lines = [
    "  what declaring cost or paid, measured by suppressing it and re-running:",
    "    unit          suppressed   declared   delta   reading",
  ];
  for (const m of measured) {
    const reading = m.delta < 0 ? "declaring COST it"
      : m.delta > 0 ? "declaring PAID"
        : "declared nothing - the control";
    lines.push(`    ${pad(m.source, 13)} ${pad(m.suppressed, 12)} ${pad(m.declared, 10)} `
      + `${pad(m.delta > 0 ? `+${m.delta}` : m.delta, 7)} ${reading}`);
  }
  return lines.join("\n");
}

File: mssp/examples/019-applicability-is-part-of-the-answer/src/DMS/report.py
"""What a person is shown.

An unmeasured reading is never rendered in the same column shape as a measured
one. `n/a` is a different kind of cell from `0`, and the report says which.
"""


def measurements(measured):
    lines = ["  unit              suppressed   declared   incentive   reading"]
    for m in measured:
        if not m["applicable"]:
            lines.append(f'  {m["source"]:<17} {"-":<12} {m["declared"]:<10} {"n/a":<11} '
                         f'NOT MEASURED: {m["reason"]}')
            continue
        reading = ("declaring PAID" if m["value"] > 0
                   else "declaring COST it" if m["value"] < 0
                   else "measured zero - both arms ran and agreed")
        value = f'+{m["value"]}' if m["value"] > 0 else str(m["value"])
        lines.append(f'  {m["source"]:<17} {m["suppressed"]:<12} {m["declared"]:<10} '
                     f'{value:<11} {reading}')
    return "\n".join(lines)


def what_a_bare_number_would_have_said(measured):
    zeros = [m for m in measured if not m["applicable"] or m["value"] == 0]
    lines = ["  what a bare number would have said:"]
    for m in zeros:
        state = "not measurable" if not m["applicable"] else "measured, and genuinely zero"
        lines.append(f'    {m["source"]:<17} 0 - and this is {state}')
    kinds = len({m["applicable"] for m in zeros})
    lines.append(f"    {len(zeros)} units, one number, {kinds} kinds of answer")
    return "\n".join(lines)


def refusal(error):
    return "\n".join([
        f"  the total refuses: {error}",
        "  A sum that skips what it could not measure reports a smaller number with the",
        "  same confidence as a complete one.",
    ])
